onnxruntime/core/session/provider_bridge_ort.cc

```diff
@@ -1419,20 +1419,72 @@ OrtTensorRTProviderOptionsV2 OrtTensorRTProviderOptionsToOrtTensorRTProviderOpti
+    } else {
+      LOGS_DEFAULT(VERBOSE) << "Invalid ep.context_embed_mode: " << embed_mode << " only 0 or 1 allowed. Set to 1.";
+    }
+    LOGS_DEFAULT(VERBOSE) << "User specified context cache embed mode: " << tensorrt_options->trt_ep_context_embed_mode;
+
+    auto context_hardware_arch_enable = (session_options->value).config_options.GetConfigOrDefault(kOrtSessionOptionEpContextHardwareArchitectureEnable, "0") != "0";
+    tensorrt_options->trt_ep_context_compute_capability_enable = context_hardware_arch_enable;
+    LOGS_DEFAULT(VERBOSE) << "User specified context hardware architecture enable: " << tensorrt_options->trt_ep_context_compute_capability_enable;
+  }
+}
+
+std::shared_ptr<IExecutionProviderFactory> TensorrtProviderFactoryCreator::Create(int device_id) {
+  return s_library_tensorrt.Get().CreateExecutionProviderFactory(device_id);
+}
+
 std::shared_ptr<IExecutionProviderFactory> TensorrtProviderFactoryCreator::Create(const OrtTensorRTProviderOptions* provider_options) {
   OrtTensorRTProviderOptionsV2 trt_options_converted = onnxruntime::OrtTensorRTProviderOptionsToOrtTensorRTProviderOptionsV2(provider_options);
   return s_library_tensorrt.Get().CreateExecutionProviderFactory(&trt_options_converted);
 }
 
+std::shared_ptr<IExecutionProviderFactory> TensorrtProviderFactoryCreator::Create(void* session_options, const OrtTensorRTProviderOptions* provider_options) {
+  OrtTensorRTProviderOptionsV2 trt_options_converted = onnxruntime::OrtTensorRTProviderOptionsToOrtTensorRTProviderOptionsV2(provider_options);
+  onnxruntime::UpdateOrtTensorRTProviderOptionsV2FromSessionOptionsConfigs(reinterpret_cast<OrtSessionOptions*>(session_options), &trt_options_converted);
+  return s_library_tensorrt.Get().CreateExecutionProviderFactory(&trt_options_converted);
+}
+
 std::shared_ptr<IExecutionProviderFactory> TensorrtProviderFactoryCreator::Create(const OrtTensorRTProviderOptionsV2* provider_options) {
   return s_library_tensorrt.Get().CreateExecutionProviderFactory(provider_options);
 }
 
+std::shared_ptr<IExecutionProviderFactory> TensorrtProviderFactoryCreator::Create(void* session_options, const OrtTensorRTProviderOptionsV2* provider_options) {
+  // We need to create a new provider options V2 object and copy from provider_options, due to the "const" object pointed by provider_options can't be modified.
+  // 
+  // Note: No need to worry about tensorrt_options being a local variable, CreateExecutionProviderFactory() in TRT EP will
+  // create a factory object that copies any provider options from tensorrt_options including "const char*" provider options.
+  OrtTensorRTProviderOptionsV2 tensorrt_options = *provider_options; // copy and assign from provider_options
+  onnxruntime::UpdateOrtTensorRTProviderOptionsV2FromSessionOptionsConfigs(reinterpret_cast<OrtSessionOptions*>(session_options), &tensorrt_options);
+  return s_library_tensorrt.Get().CreateExecutionProviderFactory(&tensorrt_options);
+}
```

Comment:
Do I misread this or is the priority in the description of this PR reversed ? This looks to me like I first update my provider options, but when appending they might be superseded by the session options. The description says that provider options supersede the session options though.

Reply:
This function is called only when "EP context" provider option is disabled and "EP context" session option is enabled.
    
    // If EP context configs are provided in session options, we need to propagate them to provider options. However,
    // if provider options already have the EP context configs provided, the configs in session options will be ignored
    // since provider options has higher priority than session options.
    if (!ep_context_cache_enabled_from_provider_options && ep_context_cache_enabled_from_sess_options) {
      factory = onnxruntime::TensorrtProviderFactoryCreator::Create(options, tensorrt_options);
    } else {
      factory = onnxruntime::TensorrtProviderFactoryCreator::Create(tensorrt_options);
    } 

so, "provider options has higher priority over session options" is still true.